### PR TITLE
Notices: Redux everything!

### PR DESCRIPTION
**Pull request by artpi**
Part of ongoing effort to clean up notices:
#888 #1094 #1105 #1313 #1364

This is how notices will look like in the future: #1415

# What is going on here?
Well, we want global notices to be in redux store.

This PR introduces second system for global notices, which relies on redux store.
Because sending notices needs the component to be connected and we have many of those, I want to have both notices systems operational at once so that we can gradually move from old to new one in a couple of PRs.
Then, we will phase out the old one.

This PR provides:
- Store for notices
- Component for displaying them
- Shows notices from `notification-settings` as an example.

# But you're doing it wrong!
I'm open fo suggestions :)

# Plan for now
- Polish and finish this. This will introduce second global notification system and 2 will be active at once. 
- in different PRs gradually switch from old notice system to new one implementing more comple use cases
- Phase out old system

## Why so many changes?
1. `global-notices` have to display notices
2. To send notices, every component sending them needs to be connected
3. To connect those components, `Layout` needs to be wrapped in Redux `Provider`
4. Because `Layout` is now wrapped in Provider, we cannot call `context.layout.setState`
5. We need to move `Layout` state to redux store.

# Testing
All around click-through is much appreciated. There are many new redux providers here and many old components are now reduxified.

The only place the notices work 'the new way' is in:
1. `http://calypso.localhost:3000/me/notifications`
2. Pick a site
3. Change notifications checkbox
4. Save
5. See the Notice 'Settings saved' is there. Awesome!

All the other notices should work also. I hope I didnt break them

### CODE REVIEW

**Comment by artpi** [comment]
I have to change PR because the base branch merged into master

**Comment by artpi** [comment]
Summary of discussion between @mcsf, @aduth and myself:
@mcsf:
I've chatted a little with @artpi and wanted to share some of my thoughts on how to do this. Basically, so much of Redux (namely, pure actions) is about declarative, data-centered code. Also, Redux middleware is a central instrument to do smart automations from inbound data/events. So the idea is to, as much as possible, reuse existing actions (that describe meaningful events, like a post being published — unlike an action like `CREATE_NOTICE`, which in a sense is an action **derived** from a real action like `PUBLISHED_POST`) using actions' `meta` subtree:

```js
// Somewhere in Calypso, someone dispatches:

{
  type: 'PUBLISHED_POST',
  postId: 42,
  meta: {
    notice: {
      text: 'Your post has been published.',
  	// defaults can be omitted:
      status: 'info',
      duration: 0,
      showDismiss: true,
      className: ''
    }
  }
}

// In the dispatcher/middleware pipeline:

const noticesMiddleware = store => next => action {
  next( action );
  if ( noticeWorthy( action ) ) {
    store.dispatch( makeNotice( action ) );
  }
}

// An event is 'notice-worthy' if its action carries `notice` meta or if
// it's an exceptional action that we care about, e.g. errors.
const noticeWorthy = action =>
  typeof action === 'object' && (
      action.meta && action.meta.notice ||
      contains( LIST_OF_SERIOUS_ACTIONS, action.type ) );

const makeNotice = action => ( dispatch, getState ) => {
  const id = ++nextId;
  const { notice } = action.meta;

  dispatch( {
    type: CREATE_NOTICE,
    id,
    action,
    notice
  } );

  if ( notice.duration )
    setTimeout( dispatch( {
      type: DELETE_NOTICE,
      id
    } ), notice.duration );
};


// In the notices reducer:

const reducer = ( state = initialState, action ) => {
  switch ( action.type ) {
    CREATE_NOTICE:
      return {
        ...state,
        [ action.id ]: action.action
      };

    DELETE_NOTICE:
      return omit( state, action.id );
  }
};
```

Me:
@mcsf 's proposition makes A LOT of sense, because it would mean moving control (displaying a notice is a controllers concern) away from view (component). 

@aduth :
>I've chatted a little with @artpi and wanted to share some of my thoughts on how to do this

@mcsf : I like this idea. It's a departure from how we use notices currently, though I think it's accurate to think of notices as being in response to other actions in the system. Some questions we can think about as we move towards something like this:

- What if I want to publish a post but not trigger a notice? Do we really want to tie data syncing concerns to the UI like this?
- Does it make sense to have a consolidated mapping of action types to notice messages, rather than attaching as meta?

@mcsf 
> * What if I want to publish a post but not trigger a notice? Do we really want to tie data syncing concerns to the UI like this?

Hm. I don't have the answer for this yet.

> * Does it make sense to have a consolidated mapping of action types to notice messages, rather than attaching as meta?

My initial reaction was to work out something like that, but as I was discussing it in-place with @seear and @ehg (who'd mumbled something about middleware) they suggested going with the one simplest approach, and stating everything in the payload seemed to satisfy that.

File: client/state/notices/actions.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import {
+	NEW_NOTICE,
+    REMOVE_NOTICE
+} from './action-types';
+
+import { uniqueId } from 'lodash';
+
+export function createNoticeAction( status, text, options = {} ) {
+	return {
+		noticeId: uniqueId(),
```
Comment (artpi) [inline client/state/notices/actions.js]: This can cause problems in de-hydration and re-hydration.
I think a good solution would be to create a helper that stores uniqueID in the store.
What do you think?

I'm leaving it as it is for now, but we need to think a bit about these kind of random pieces of data.

File: client/state/index.js [hunk 2]
```diff
@@ -10,6 +10,7 @@ import { createStore, applyMiddleware, combineReducers } from 'redux';
 import sharing from './sharing/reducer';
 import sites from './sites/reducer';
 import ui from './ui/reducer';
+import notices from './notices/reducers';
```
Comment (artpi) [inline client/state/index.js]: Do we want notices in the `/ui` subtree?

File: client/state/notices/actionCreators.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { createNoticeAction, removeNoticeAction } from './actions';
+
+export default function( dispatch ) {
+	function createNotice( type, text, options ) {
+		var action = createNoticeAction( type, text, options );
+
+		if ( action.duration > 0 ) {
+			setTimeout( () => {
```
Comment (artpi) [inline client/state/notices/actionCreators.js]: @gziolo said: in #1041 :
> Another param that can cause troubles is duration. Though in this case we can trigger action using component lifecycle methods. I'm not sure if that's good practice, but we could use componentDidMount and set timeout which would trigger action after given duration.

I know this approach here does not survive dehydration.
We will need to approach it somehow, but I dont want to get component involved at this point.

File: client/boot/index.js [hunk 4]
```diff
@@ -177,12 +178,14 @@ function boot() {
...
 
-		layout = React.render(
-			React.createElement( LoggedOutLayout ),
-			document.getElementById( 'wpcom' )
-		);
+		layoutComponentCreator = () => React.createElement( LoggedOutLayout );
 	}
 
+	layout = React.render(
+		React.createElement( ReduxProvider, { store: reduxStore }, layoutComponentCreator ),
```
Comment (aduth) [inline client/boot/index.js]: Just a heads-up that I had found in the course of upgrading to React 0.14 in #1498 that this will need to pass the child directly, not as function, even without the upgrade to `react-redux` 4.x.

File: client/state/notices/actionCreators.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { createNoticeAction, removeNoticeAction } from './actions';
+
+export default function( dispatch ) {
```
Comment (aduth) [inline client/state/notices/actionCreators.js]: I'm not sure I like the naming of this file. First, it doesn't adhere to our naming convention of using dash delimiters, not camel-case. More importantly, the functions in `actions.js` are truly action creators themselves. The function here is something like an action creator factory (action creator creator? :smile: ). Personally, I'd prefer that we simplify to a single `createNotice` asynchronous action creator in `actions.js`, with bound convenience action creators. If you wanted, you could still choose to split off the synchronous action creator, though I don't see the benefit of doing so.

```jsx
// client/state/notices/actions.js

export function createNotice( status, text, options ) {
	return function( dispatch ) {
		dispatch( { /* ... */ } );

		setTimeout( () => {
			dispatch( { /* ... */ } );
		}, options.duration );
	}
}

export const createSuccessNotice = createNotice.bind( null, 'is-success' );
```

File: client/me/controller.js [hunk 6]
```diff
@@ -192,15 +193,17 @@ export default {
...
-			React.createElement( NotificationsComponent,
-				{
-					user: user,
-					userSettings: userSettings,
-					blogs: sites,
-					devices: devices,
-					path: context.path
-				}
-			),
+			React.createElement( Provider, { store: context.store }, () => {
```
Comment (drewblaisdell) [inline client/me/controller.js]: If the root layout component is wrapped in `<Provider />`, why do we need to wrap `NotificationsComponent` in it here? Would upgrading to 0.14 obviate this?

Comment (aduth) [inline client/me/controller.js]: >If the root layout component is wrapped in `<Provider />`, why do we need to wrap `NotificationsComponent` in it here? Would upgrading to 0.14 obviate this?

It's unfortunately needed in both places, since there's technically two separate React render points. React 0.14 won't help here, though I think part of the idea with future plans for a global `<App />` component is to have a single render hierarchy.

File: client/state/notices/actionCreators.js (hunk 5 above)
Comment (artpi) [inline client/state/notices/actionCreators.js]: Will change, thanks!

File: client/me/controller.js (hunk 6 above)
Comment (artpi) [inline client/me/controller.js]: And if we adopt `redux-router` that will also change things.
But single point of render would be awesome

File: client/boot/index.js (hunk 4 above)
Comment (artpi) [inline client/boot/index.js]: I will wait for you to merge. I want you to merge your PR ASAP

File: client/me/controller.js (hunk 6 above)
Comment (ehg) [inline client/me/controller.js]: > But single point of render would be awesome

FWIW, I was having a play with doing this with the existing `page.js` setup  in https://github.com/Automattic/wp-calypso/pull/1186

File: client/state/index.js (hunk 2 above)
Comment (artpi) [inline client/state/index.js]: So yeah, it's now in `/ui` subtree

File: client/boot/index.js [hunk 7]
```diff
@@ -162,13 +163,14 @@ function boot() {
 
 		// Create layout instance with current user prop
 		Layout = require( 'layout' );
-		layout = ReactDom.render( React.createElement( Layout, {
+
+		layoutComponent = React.createElement( Layout, {
```
Comment (aduth) [inline client/boot/index.js]: Really nit-picking at this point, but terminology here may be inaccurate, as this is assigning an element, not a component (similar to but not quite the same as the distinction between a class and instance of the class).

https://facebook.github.io/react/blog/2014/10/14/introducing-react-elements.html

```jsx
layoutElement = React.createElement( /* ... */ );
```

File: client/components/global-notices/index.jsx [hunk 8]
```diff
@@ -12,11 +12,13 @@ import Notice from 'components/notice';
...
 import observe from 'lib/mixins/data-observe';
-import DeleteSiteNotices from './delete-site-notices';
+import DeleteSiteNotices from 'notices/delete-site-notices';
+import { connect } from 'react-redux';
+import { noticesMapDispatchToProps } from 'state/ui/notices/actions'
 
 const debug = debugModule( 'calypso:notices' );
 
-export default React.createClass( {
+const NoticesList =  React.createClass( {
```
Comment (aduth) [inline client/components/global-notices/index.jsx]: Additional space after the `=` to be removed.

Consider using an ESLint plugin for your preferred editor if you're not already, as they often make these errors easy to spot.

image

File: client/components/global-notices/index.jsx [hunk 9]
```diff
@@ -79,7 +81,7 @@ export default React.createClass( {
 
 	render() {
 		const noticesRaw = this.props.notices[ this.props.id ] || [];
-		const noticesList = noticesRaw.map( function( notice, index ) {
+		var noticesList = noticesRaw.map( function( notice, index ) {
```
Comment (aduth) [inline client/components/global-notices/index.jsx]: Why was this changed from `const` to `var` ?

File: client/components/global-notices/index.jsx (hunk 8 above)
Comment (artpi) [inline client/components/global-notices/index.jsx]: I'm using it, this is merge conflict poorly resolved :/
Thanks!

File: client/boot/index.js (hunk 7 above)
Comment (artpi) [inline client/boot/index.js]: Will change! Thanks!

File: client/layout/index.jsx [hunk 10]
```diff
@@ -123,3 +117,13 @@ module.exports = React.createClass( {
 		);
 	}
 } );
+
+export default connect(
+	( state ) => {
+		return {
```
Comment (blowery) [inline client/layout/index.jsx]: if you did `const { isLoading, section, noSidebar } = state.ui;` you could then simplify the object literal to `{ isLoading, section, noSidebar }`.

No necessary, I just find it easier to read.

File: client/state/ui/notices/actions.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+export function removeNoticeAction( noticeId ) {
+	return {
+		noticeId: noticeId,
+		type: REMOVE_NOTICE
+	};
+}
+
+export function noticesMapDispatchToProps( dispatch ) {
```
Comment (aduth) [inline client/state/ui/notices/actions.js]: `mapDispatchToProps` functions feel out of place in `actions.js`, since they're very much tied to React in that they define a component's props bound to action creators. In this case, if the desire is to have a named function, I think it's reasonable to define in the same file as the component, directly above the call to `connect`.

```jsx
const NoticesList = React.createClass( { /* ... */ } );

function mapStateToProps( state ) {
	// ...
}

function mapDispatchToProps( dispatch ) {
	// ...
}

export default connect( mapStateToProps, mapDispatchToProps )( NoticesList );
```

While I could see some re-use between these mapping functions, I don't think it's a convention we should promote, as it might lead us to creating very generic mapping connectors that pass complex handlers intended to handle a variety of use-cases. Instead, I think `connect` arguments should be specialized to the particular component its wrapping.

Comment (aduth) [inline client/state/ui/notices/actions.js]: Hmm, I guess this goes back to the desire to have easy-to-use interactions for notices across the entire application? In that case, the potential for re-use is more legitimate. I still don't think it's appropriate for `actions`, or even to be defined in the `state` directory at all. Maybe this could be included as a file in `client/lib/mixins` ?

File: client/devdocs/controller.js [hunk 12]
```diff
@@ -52,7 +53,8 @@ var devdocs = {
 				false );
 		}
 
-		context.layout.setState( {
+		context.store.dispatch( {
```
Comment (aduth) [inline client/devdocs/controller.js]: For usability's sake, why don't we create an action creator here?

```jsx
context.store.dispatch( setSection( 'devdocs' ) );
```

File: client/me/controller.js (hunk 6 above)
Comment (aduth) [inline client/me/controller.js]: This still needs to be updated to not render a function child, but instead the element directly.

File: client/me/controller.js [hunk 13]
```diff
@@ -193,15 +196,17 @@ export default {
...
+				return React.createElement( NotificationsComponent,
```
Comment (aduth) [inline client/me/controller.js]: To keep indenting lower, consider moving the opening and closing braces:

```jsx
React.createElement( NotificationsComponent, {
	user,
	userSettings,
	blogs: sites,
	devices,
	path: context.path
} )
```

File: client/me/notification-settings/index.jsx [hunk 14]
```diff
@@ -74,3 +75,9 @@ export default React.createClass( {
 	}
 } );
 
+export default connect(
+	() => {
```
Comment (aduth) [inline client/me/notification-settings/index.jsx]: I believe you can pass `null` if you don't need to pass props from state, which is much more performant as `connect` will not need to re-render on state change.

File: client/state/ui/notices/actions.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+	REMOVE_NOTICE
+} from 'state/action-types';
+
+import { uniqueId } from 'lodash';
+
+export function createNoticeAction( status, text, options = {} ) {
+	return {
+		noticeId: uniqueId(),
+		duration: options.duration,
+		showDismiss: ( typeof options.showDismiss === 'boolean' ? options.showDismiss : true ),
```
Comment (aduth) [inline client/state/ui/notices/actions.js]: Rather than type-checking, if the goal is to default to `true` if not defined, we could check for the presence of the key:

```jsx
showDismiss: 'showDismiss' in options ? options.showDismiss : true
```

File: client/state/ui/notices/reducer.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Internal dependencies
+ */
+import {
+	NEW_NOTICE,
+	REMOVE_NOTICE
+} from 'state/action-types';
+
+/**
+ * Tracks all known site objects, indexed by site ID.
```
Comment (aduth) [inline client/state/ui/notices/reducer.js]: I don't think this comment is applicable :smile:

File: client/state/ui/notices/reducer.js [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function items( state = [], action ) {
```
Comment (aduth) [inline client/state/ui/notices/reducer.js]: Should we index this by ID? Would make removing and lookup simpler and potentially more performant (avoiding the need to iterate over the array).

```jsx

export function items( state = {}, action ) {
	switch ( action.type ) {
		case NEW_NOTICE:
			state = Object.assign( {}, state, {
				[ action.notice.id ]: action.notice
			} );
			break;
		case REMOVE_NOTICE:
			state = omit( state, action.notice.id );
			break;
	}

	return state;
}
```

File: client/state/ui/notices/actions.js [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+		type: NEW_NOTICE,
```
Comment (aduth) [inline client/state/ui/notices/actions.js]: The way we're tracking state below will include the `type` key in the state tree, which is undesirable. Instead, consider nesting the notice object in `notice` (or `payload`, or `data`):

```jsx
export function createNoticeAction( status, text, options = {} ) {
	return {
		type: NEW_NOTICE,
		notice: {
			id: uniqueId(),
			duration: options.duration,
			showDismiss: ( typeof options.showDismiss === 'boolean' ? options.showDismiss : true ),
			status: status,
			text: text
		}
	};
}
```

File: client/state/ui/notices/reducer.js [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+	switch ( action.type ) {
+		case NEW_NOTICE:
+			state = [ action, ...state ];
+			break;
+		case REMOVE_NOTICE:
+			state = filter( state, ( notice ) => ( notice.noticeId !== action.noticeId ) );
```
Comment (aduth) [inline client/state/ui/notices/reducer.js]: Also, while Lodash's functions have sometimes been shown to be faster than their native equivalents, you don't need a module for `filter`:

```jsx
state = state.filter( ( notice ) => ( notice.noticeId !== action.noticeId ) );
```

File: client/state/ui/reducer.js [hunk 20]
```diff
@@ -25,6 +26,31 @@ export function selectedSite( state = null, action ) {
 	return state;
 }
 
+export function section( state = false, action ) {
+	if ( action.type === SET_SECTION && action.section !== undefined ) {
```
Comment (aduth) [inline client/state/ui/reducer.js]: Do we need to check against `undefined` ? I worry that excessive validation hurts readability, when we should have the expectation that the actions be called with valid parameters.

Comment (artpi) [inline client/state/ui/reducer.js]: Because we have 3 reducers for the same `SET_SECTION` action, they are invoked only when user is feeding a isLoading / noSidebar / section
While I agree that they should be more granulated reducers, I honestly cannot tell from the old code calls that were `context.layout.setState ({section:, noSidebar, isLoading}) ` which call was intended.
I built in this safe-check to only reduce what was meant.
I opt to merge this and figure it out later.

File: client/state/ui/notices/reducer.js (hunk 17 above)
Comment (artpi) [inline client/state/ui/notices/reducer.js]: I don't expect to have that many notices shown at once for it to really make a difference.
And there would be an additional effort to keep them sorted chronologically while displaying. 
> The for...in statement iterates over the enumerable properties of an object, in arbitrary order

File: client/boot/index.js (hunk 7 above)
Comment (artpi) [inline client/boot/index.js]: Changed in c60c45a39508ff76fb992245199309d3e03aaf6b

File: client/components/global-notices/index.jsx (hunk 8 above)
Comment (artpi) [inline client/components/global-notices/index.jsx]: Changed in c60c45a39508ff76fb992245199309d3e03aaf6b

File: client/components/global-notices/index.jsx (hunk 9 above)
Comment (artpi) [inline client/components/global-notices/index.jsx]: Changed in 7d9789696a5e799ab16c0df6efc362caaa8bf4d0

File: client/state/ui/notices/reducer.js (hunk 19 above)
Comment (artpi) [inline client/state/ui/notices/reducer.js]: Changed in 9adb1e68423db3a55d14edcb7153a4145caa0aba

File: client/me/notification-settings/index.jsx (hunk 14 above)
Comment (artpi) [inline client/me/notification-settings/index.jsx]: Changed in c37899d0efc005ee1b6d12b2dfb13e0fb82be257

File: client/me/controller.js (hunk 6 above)
Comment (artpi) [inline client/me/controller.js]: Changed in 767babc9c4a3b786edb233c67b7ddc69eb478ad1

Comment (artpi) [inline client/me/controller.js]: hanged in 767babc9c4a3b786edb233c67b7ddc69eb478ad1

File: client/me/controller.js (hunk 13 above)
Comment (artpi) [inline client/me/controller.js]: Changed in 767babc9c4a3b786edb233c67b7ddc69eb478ad1

File: client/state/ui/notices/actions.js (hunk 11 above)
Comment (artpi) [inline client/state/ui/notices/actions.js]: But then again it is not a mixin and there is no way to turn it into one.
I completely agree, but I have no place better for it.
And we need this helper, it will have ~20 lines of code we would have to copy across 60+ files

File: client/state/ui/notices/actions.js (hunk 18 above)
Comment (artpi) [inline client/state/ui/notices/actions.js]: I agree!
Changed in c7419d4b1f20863a82d14b696f77bb61f379681b

File: client/state/ui/notices/reducer.js (hunk 16 above)
Comment (artpi) [inline client/state/ui/notices/reducer.js]: Removed in c7419d4b1f20863a82d14b696f77bb61f379681b

File: client/state/ui/notices/actions.js (hunk 11 above)
Comment (aduth) [inline client/state/ui/notices/actions.js]: >But then again it is not a mixin and there is no way to turn it into one.

Depends on the definition of mixin :smiley: It does appear that the majority of what we have there are "React mixins", though the [`emitter` mixin](https://github.com/Automattic/wp-calypso/tree/master/shared/lib/mixins/emitter) is one exception. An alternative is to have a directory in `lib` containing these types of helpers.

>And we need this helper, it will have ~20 lines of code we would have to copy across 60+ files

That seems a bit of a stretch, and I still question whether we need these helpers. Sure, what we have here is 20 lines, but I still think this could be simplified to a one-line function in the `connect` binding:

```jsx
function Button( { createSuccessNotice } ) {
	return (
		<button onClick={ () => createSuccessNotice( 'Great success!' ) }>
			Create Success
		</button>
	);
}

export default connect( null, ( dispatch ) => {
	return bindActionCreators( { createSuccessNotice }, dispatch );
} )( Button );
```

File: client/state/ui/notices/reducer.js (hunk 17 above)
Comment (aduth) [inline client/state/ui/notices/reducer.js]: >And there would be an additional effort to keep them sorted chronologically while displaying.

Fair point :+1:

**Comment by aduth** [comment]
Looks like this needs a rebase to resolve merge conflicts (and a few commits worth squashing).

File: client/state/ui/notices/actions.js (hunk 11 above)
Comment (artpi) [inline client/state/ui/notices/actions.js]: Well I somehow missed `bindActionCreators`! And yeah, that solves a lot :)
Will go this route.

File: client/state/ui/notices/actions.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
+/**
```
Comment (mtias) [inline client/state/ui/notices/actions.js]: I feel like `notices` should be at root level, not under `ui`. They are only incidentally ui state because we choose to render them in layout, not because it's directly state of the UI. Specially given we may just tied them as a sort of logger for other actions.

File: client/state/ui/reducer.js [hunk 22]
```diff
@@ -25,6 +26,31 @@ export function selectedSite( state = null, action ) {
...
+		state = action.section;
+	}
+	return state;
+}
+
+export function noSidebar( state = false, action ) {
```
Comment (mtias) [inline client/state/ui/reducer.js]: Given we are moving things around, I think we should rename this to be the opposite `sidebar: true`.

File: server/bundler/loader.js [hunk 23]
```diff
@@ -63,18 +63,18 @@ function splitTemplate( path, module, chunkName ) {
 		'		layoutFocus.next();',
 		'		return next();',
 		'	}',
-		'	context.layout.setState( { isLoading: true } );',
+		'	context.store.dispatch( { type: "SET_SECTION", isLoading: true } );',
```
Comment (mtias) [inline server/bundler/loader.js]: This is much better, and would help when we read from `sections.js`

File: client/devdocs/controller.js [hunk 24]
```diff
@@ -72,10 +70,7 @@ var devdocs = {
 	 * Controller for single developer document
 	 */
 	singleDoc: function( context ) {
-		context.layout.setState( {
-			section: 'devdocs',
-			noSidebar: false
-		} );
+		context.store.dispatch( setSection( 'devdocs' ) );
```
Comment (mtias) [inline client/devdocs/controller.js]: Does this work ok without declaring the sidebar state attribute?

Comment (artpi) [inline client/devdocs/controller.js]: I made `noSidebar: false` the default.

File: client/state/ui/reducer.js (hunk 22 above)
Comment (artpi) [inline client/state/ui/reducer.js]: Makes sense

File: client/state/ui/notices/actions.js (hunk 11 above)
Comment (artpi) [inline client/state/ui/notices/actions.js]: Ok, I remember now why I didn't use `bindActionCreators`.
I need `dispatch` in closure there to set callback for timeout for the notice to disappear. 
Every other solution does not give me a direct access to `dispatch`.
If I use `bindActionCreators`, I would have to create some ticker that keeps track of notices and `disappears` them. This ticker would have to be in controller (layout), because this is the only place that has direct access to `context.store.dispatch`

This can be a good solution, but definitely for a different PR.
For the time beeing, I can move this helper, but as I said - I have no better place.
`client/lib/redux-store-helpers` ?

File: client/state/ui/reducer.js (hunk 22 above)
Comment (artpi) [inline client/state/ui/reducer.js]: I made `hasSidebar: true`

File: client/state/ui/notices/actions.js (hunk 21 above)
Comment (artpi) [inline client/state/ui/notices/actions.js]: I kind of agree with @aduth that this is part of UI and probably for now lets leave it in `ui`
And I dont think we should notices as loggers for other actions, I would like to see it the other way around - notices beeing a by-product of other actions

File: client/layout/index.jsx (hunk 10 above)
Comment (gziolo) [inline client/layout/index.jsx]: Another alternative:
```js
pick( state.ui, [ 'hasSidebar', 'isLoading', 'section' ] );
```
:)

File: client/state/ui/notices/actions.js (hunk 21 above)
Comment (mtias) [inline client/state/ui/notices/actions.js]: I don't quite like UI as a catch all. There's also the fact that we'll be showing more than one notice at a time, but probably not all, so there's not even a 1-to-1 correlation between the state and what we display.

File: client/state/ui/notices/actions.js (hunk 11 above)
Comment (artpi) [inline client/state/ui/notices/actions.js]: I used `thunk` and it is now nice and proper!
Changed in 8206a4e784ceeee530c7b7e6288913b0f0f935e7
thunk + bindActionCreators produce a nice solution.

Comment (aduth) [inline client/state/ui/notices/actions.js]: >thunk + bindActionCreators produce a nice solution.

Agree! The thunk middleware is a good fit for these types of asynchronous actions.

File: client/state/ui/notices/actions.js (hunk 21 above)
Comment (artpi) [inline client/state/ui/notices/actions.js]: Moved to top level in 63149e0db7ec941336a183cd7c2400c2c78c4d10

File: client/layout/index.jsx (hunk 10 above)
Comment (rralian) [inline client/layout/index.jsx]: Either of those are nice, but I like the destructuring approach a little more.

File: client/layout/logged-out.jsx [hunk 25]
```diff
@@ -41,3 +36,13 @@ module.exports = React.createClass( {
 	}
 
 } );
+
+export default connect(
+	( state ) => {
+		return {
+			isLoading: state.ui.isLoading,
```
Comment (rralian) [inline client/layout/logged-out.jsx]: I don't think you need isLoading here, do you?

File: client/state/notices/actions.js [hunk 26]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Internal dependencies
+ */
+import {
+	NEW_NOTICE,
+	REMOVE_NOTICE
+} from 'state/action-types';
+
+import { uniqueId } from 'lodash';
```
Comment (rralian) [inline client/state/notices/actions.js]: Will any of this run on the server? Because I believe `uniqueId` will then generate unique IDs across all sessions... so I think we could potentially run into memory issues?

Comment (artpi) [inline client/state/notices/actions.js]: That is correct, they will be unique across.
Memory wont probably be a problem, because max integer is `9007199254740991` so that's a lot of notices and I suspect we will run into many other leaks before this becomes an issue.
That being said, the problem here is state (de)hydration

Let's say we are dehydrating notices and next id is 4.
When we reload and  re-hydrate, the next id is 1 and we already have notice with id 1 in store.

So, I intend to make a uniqueID in global store with state saving
But this is  probably a matter for another PR :)
Probably can started on that right away.

Comment (gziolo) [inline client/state/notices/actions.js]: You can pass env as prefix param for `uniqueId` and issue solved :)

Comment (artpi) [inline client/state/notices/actions.js]: @gziolo, this is not a solution, because you would need globally uniqueId of the env to pass to uniqueId :)

But, behold:
https://github.com/Automattic/wp-calypso/pull/1793

File: client/layout/index.jsx (hunk 10 above)
Comment (artpi) [inline client/layout/index.jsx]: Changed!

File: client/layout/logged-out.jsx (hunk 25 above)
Comment (artpi) [inline client/layout/logged-out.jsx]: Changed!

File: client/components/global-notices/index.jsx [hunk 27]
```diff
@@ -12,11 +12,14 @@ import Notice from 'components/notice';
...
-import DeleteSiteNotices from './delete-site-notices';
+import DeleteSiteNotices from 'notices/delete-site-notices';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import { removeNotice } from 'state/notices/actions'
 
 const debug = debugModule( 'calypso:notices' );
 
-export default React.createClass( {
+const NoticesList = React.createClass( {
```
Comment (aduth) [inline client/components/global-notices/index.jsx]: Should we change the declared name of the component and its class to `<GlobalNotices />` and `.global-notices` respectively, given the new file location?

File: client/layout/logged-out.jsx [hunk 28]
```diff
@@ -2,36 +2,31 @@
...
-			noSidebar: false
-		};
-	},
-
 	render: function() {
-		var sectionClass = this.state.section ? ' is-section-' + this.state.section : '',
+		var sectionClass = this.props.section ? ' is-section-' + this.props.section : '',
 			classes = classNames( 'wp', sectionClass, {
-				'has-no-sidebar': this.state.noSidebar
+				'has-no-sidebar': !this.props.hasSidebar
```
Comment (aduth) [inline client/layout/logged-out.jsx]: Nit-picking: Should be a space after the `!` ([this'll eventually be part of a potential future custom ESLint rule](https://github.com/Automattic/wp-calypso/blob/8171f7a6002489800fa68be0ba5c88af5adde634/.eslintrc#L81-L82)).

File: client/me/controller.js [hunk 29]
```diff
@@ -303,7 +302,7 @@ export default {
 
 		if ( isWelcome ) {
 			ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
-			context.layout.setState( { noSidebar: true } );
+			context.store.dispatch( setSection( null, { hasSidebar: false } ) );
```
Comment (aduth) [inline client/me/controller.js]: This makes me wonder whether we should have separate actions for `sidebar` and `isLoading`, since they're not always tied to the section as is the case here.

**Comment by aduth** [comment]
This is looking to be in a really good state. A few remaining questions, but otherwise I think we'll soon be mergeable.

File: client/components/global-notices/index.jsx (hunk 27 above)
Comment (artpi) [inline client/components/global-notices/index.jsx]: @mtias is rewriting this file in #1415 and I dont want to put much effort into it :)
It will be completely overhauled based on those 2 PRs

File: client/me/controller.js (hunk 29 above)
Comment (artpi) [inline client/me/controller.js]: While I agree this can be the case for isLoading, `hasSidebar` is very strongly logically connected to a section it should acompany.
This instance here is one of 3 in entire code where sidebar is set without pointing a section and is - in my opinion - a hack and not entirely clear.
Thats why I didnt set seperate `hasSidebar` action in order to not encourage its usage.

When it comes to `isLoading`, it would be helpful, but code dispatching `isLoading` param is generated dynamically on the server in `sections.js` and I dispatched actions there explicitly through `{ type: SET_SECTION, isLoading: true}`, without actionCreator.
Because that code is created on the server, I didnt want to inlude additional dependencies, because it would be hard to figure out where the action creator came from while reading the soure.

File: client/layout/logged-out.jsx (hunk 28 above)
Comment (artpi) [inline client/layout/logged-out.jsx]: Fixed

**Comment by aduth** [comment]
:+1: from me

**Comment by aduth** [comment]
Probably won't break anything, but you should rebase against latest master to account for #1867.

**Comment by drewblaisdell** [comment]
It looks like this PR introduced a couple issues in checkout and `/accept-invites`, because of a typo when calling the dispatcher. A PR to fix these now exists here: https://github.com/Automattic/wp-calypso/pull/1941

File: client/components/overlay/overlay.jsx [hunk 30]
```diff
@@ -100,7 +100,7 @@ module.exports = React.createClass({
 				</TitleData>
 
 				<div className="wp-content" ref="overlayInnerContent">
-					<NoticesList id="overlay-notices" notices={ notices.list }/>
+					<GlobalNotices id="overlay-notices" notices={ notices.list }/>
```
Comment (drewblaisdell) [inline client/components/overlay/overlay.jsx]: Also, one more thing: `Overlay` isn't part of `Layout` in this case, so `GlobalNotices` didn't have access to the global state here. This was fixed in https://github.com/Automattic/wp-calypso/pull/1944

File: client/state/notices/actions.js (hunk 26 above)
Comment (aduth) [inline client/state/notices/actions.js]: We should not be importing from Lodash this way, as it increases each bundle size by about ~500kb. See #735 for a detailed explanation.
